README.md
Event Registry is a Python package that can be used to easily access the news data available in [Event Registry](http://eventregistry.org/) through the API. The package can be used to query for articles or events by filtering using a large set of filters, like keywords, concepts, topics, sources, sentiment, date, etc. Details about the News API are available on the [landing page of the product](https://newsapi.ai/).

## Installation

Event Registry package can be installed using Python's pip installer. In the command line, simply type:

    pip install eventregistry

and the package should be installed. Alternatively, you can also clone the package from the [GitHub repository](https://github.com/EventRegistry/event-registry-python). After cloning it, open the command line and run:

    python setup.py install

### Validating installation

To ensure the package has been properly installed run python and type:

```python
import eventregistry
```

If you don't get any error messages, then your installation has been successful.

### Updating the package

As features are added to the package you will need at some point to update it. In case you have downloaded the package from GitHub simply do a `git pull`. If you have installed it using the `pip` command, then simply run

	pip install eventregistry --upgrade

### Authentication and API key

When making queries to Event Registry you will have to use an API key that you can obtain for free. The details on how to obtain and use the key are described in the [Authorization](../../wiki/EventRegistry-class#authorization) section.

## Four simple examples to get you interested

**Print a list of recently articles or blog posts from *US based sources* *with positive sentiment* mentioning phrases *"George Clooney"* or *"Sandra Bullock"***

```python
from eventregistry import *
er = EventRegistry(apiKey = YOUR_API_KEY)

# get the USA URI
usUri = er.getLocationUri("USA")    # = http://en.wikipedia.org/wiki/United_States

q = QueryArticlesIter(
    keywords = QueryItems.OR(["George Clooney", "Sandra Bullock"]),
    minSentiment = 0.4,
    sourceLocationUri = usUri,
    dataType = ["news", "blog"])

# obtain at most 500 newest articles or blog posts, remove maxItems to get all
for art in q.execQuery(er, sortBy = "date", maxItems = 500):
    print(art)
```

**Print a list of most relevant *business* articles from the last month related to *Microsoft* or *Google*. The articles should be in any language (including Chinese, Arabic, ...)**

```python
from eventregistry import *
# allowUseOfArchive=False will allow us to search only over the last month of data
er = EventRegistry(apiKey = YOUR_API_KEY, allowUseOfArchive=False)

# get the URIs for the companies and the category
microsoftUri = er.getConceptUri("Microsoft")    # = http://en.wikipedia.org/wiki/Microsoft
googleUri = er.getConceptUri("Google")          # = http://en.wikipedia.org/wiki/Google
businessUri = er.getCategoryUri("news business")    # = news/Business

q = QueryArticlesIter(
    conceptUri = QueryItems.OR([microsoftUri, googleUri]),
    categoryUri = businessUri)

# obtain at most 500 newest articles, remove maxItems to get all
for art in q.execQuery(er, sortBy = "date", maxItems = 500):
    print(art)
```


**Search for latest events related to Star Wars**

```python
from eventregistry import *
er = EventRegistry(apiKey = YOUR_API_KEY)

q = QueryEvents(keywords = "Star Wars")
q.setRequestedResult(RequestEventsInfo(sortBy = "date", count = 50))   # request event details for latest 50 events

# get the full list of 50 events at once
print(er.execQuery(q))
```

**Search for articles that (a) mention immigration, (b) are related to business, and (c) were published by news sources located in New York City**

```python
from eventregistry import *
er = EventRegistry(apiKey = YOUR_API_KEY)

q = QueryArticlesIter(
    # here we don't use keywords so we will also get articles that mention immigration using various synonyms
    conceptUri = er.getConceptUri("immigration"),
    categoryUri = er.getCategoryUri("business"),
    sourceLocationUri = er.getLocationUri("New York City"))

# obtain 500 articles that have were shared the most on social media
for art in q.execQuery(er, sortBy = "socialScore", maxItems = 500):
    print(art)
```

**What are the currently trending topics**

```python
from eventregistry import *
er = EventRegistry(apiKey = YOUR_API_KEY)

# top 10 trending concepts in the news
q = GetTrendingConcepts(source = "news", count = 10)
print(er.execQuery(q))
```

## Learning from examples

We believe that it's easiest to learn how to use our service by looking at examples. For this reason, we have prepared examples of various most used features. View the examples grouped by main search actions:

[View examples of searching for articles](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/QueryArticlesExamples.py)

[View examples of searching for events](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/QueryEventsExamples.py)

[View examples of obtaining information about an individual event](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/QueryEventExamples.py)

[Examples of how to obtain the full feed of articles](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/FeedOfNewArticlesExamples.py)

[Examples of how to obtain the full feed of events](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/FeedOfNewEventsExamples.py)

## Play with interactive Jupyter notebook

To interactively learn about how to use the SDK, see examples of use, see how to get extra meta-data properties, and more, please open [this Binder](https://mybinder.org/v2/gh/EventRegistry/event-registry-python-intro/master). You'll be able to view and modify the examples.

## Where to next?

**[Terminology](../../wiki/Terminology)**. There are numerous terms in the Event Registry that you will constantly see. If you don't know what we mean by an *event*, *story*, *concept* or *category*, you should definitely check this page first.

**[Learn about `EventRegistry` class](../../wiki/Eventregistry-class)**. You will need to use the `EventRegistry` class whenever you will want to interact with Event Registry so you should learn about it.

**[Details about articles/events/concepts/categories/... that we can provide](../../wiki/ReturnInfo-class)**. When you will be requesting information about events, articles, concepts, and other things, what details can you ask for each of these?

**[Querying events](../../wiki/Searching-for-events)**. Check this page if you are interested in searching for events that match various search criteria, such as relevant concepts, keywords, date, location or others.

**[Querying articles](../../wiki/Searching-for-articles)**. Read if you want to search for articles based on the publisher's URL, article date, mentioned concepts or others.

**[Trends](../../wiki/Trends)**. Are you interested in finding which concepts are currently trending the most in the news? Maybe which movie actor is most popular in social media? How about trending of various news categories?

**[Articles and events shared the most on social media](../../wiki/Social-shares)**. Do you want to get the list of articles that have been shared the most on Facebook and Twitter on a particular date? What about the most relevant event based on shares on social media?

## Data access and usage restrictions

Event Registry is a commercial service but it allows also unsubscribed users to perform a certain number of operations. Non-paying users are not allowed to use the obtained data for any commercial purposes (see the details on our [Terms of Service page](http://newsapi.ai/terms)) and have access to only last 30 days of content. In order to avoid these restrictions please contact us about the [available plans](http://newsapi.ai/plans).

CHANGELOG.md
# Change Log

## [v9.1]() (2023-06-23)

**Added**
- added `keywordSearchMode` parameter that can be used in `QueryArticles`, `QueryArticlesIter`, `QueryEvents`, `QueryEventsIter` and `QueryEvent` constructors.
- added `keywordSearchMode` parameter to the advanced query language


**Updated**
- types of parameters in the method calls
- updated several code example files



## [v9.0]() (2023-05-15)

**Added**
- added use of the typing module. All parameters in the method calls use typing support to make it easier to understand what type is expected.
- added autosuggest methods `suggestEventTypes`, `suggestIndustries`, `getSdgUris`, `getSasbUris` - all to be used only when querying mentions
- 

**Updated**
- `QueryArticles` class. Added filters `authorsFilter`, `videosFilter`, `linksFilter`
- `QueryMentions` class. Added several filters: `industryUri`, `sdgUri`, `sasbUri`, `esgUri`, `minSentenceIndex`, `maxSentenceIndex`, `showDuplicates`
- updated several code example files


## [v8.12]() (2022-03-11)

**Updated**
- changed the logger. No longer printing out to console information about the host, etc
- removed two parameters from `EventRegistry` constructor:
	- `logging` - if you wish to log requests to file, please call now `EventRegistry.setLogging(True)`
	- `printHostInfo` - if you wish to see the host info, call `logging.getLogger("eventregistry").setLevel(logging.DEBUG)`
- `checkVersion()` is no longer automatically called in `EventRegistry` constructor. If you wish to see if there is an updated version of the module, please call the method manually.


## [v8.11]() (2022-01-05)

**Updated**
- the `QueryEventArticlesIter` class will be significantly faster in returning the articles when you specify the language filter

## [v8.10]() (2021-08-16)

**Added**

- added file `QueryMentions` that can be used to query mentions of specific event types. The class is currently in beta and not available to users unless they have permissions to use this endpoint. The classes `QueryMentions` and `QueryMentionsIter` can be used in the same way as classes for querying articles and events, except that some query parameters are addded and some removed. Examples for the classes were also added.

**Updated**
- When using method `initWithComplexQuery` we now check if the provided json is valid json object and report error in case it is not
- some bugs that appeared in edge cases when querying Event Registry using `EventRegistry` class have been fixed.

## [v8.9]() (2020-10-21)

**Added**

- added `required` and `excluded` optional parameters to the TopicPage class to methods `addConcept`, `addKeyword`, `addCategory`, `addSource`, `addSourceLocation` and `addSourceGroup`.
- exceptions are now raised by the SDK library in case of status codes 204, 400, 401 and 403. Previously, the requests were repeated if status code was not 200. In  case of these status codes it however doesn't make sense to simply repeat the requests as the query will for sure return the same exception. The meaning of the above status codes is described on the [documentation page](https://eventregistry.org/documentation?tab=introduction).


## [v8.8]() (2020-08-20)

**Added**

- added filters `reportingDateStart` and `reportingDateEnd` to `QueryEvents` constructor. You can use these dates to find the events where the average date of reporting matches this date range. Note that `dateStart` and `dateEnd` will return events based on when they (supposedly) happened and it can be quite different than the reporting dates.
- added parameters `updatesAfterNewsUri`, `updatesafterBlogUri`, `updatesAfterPrUri` to the `RequestArticlesRecentActivity` constructor. Can be used to make sure that absolutely no article is missed when using the `QueryArticles` with `RequestArticlesRecentActivity` requested result. See [this file](https://github.com/EventRegistry/event-registry-python/blob/master/eventregistry/examples/FeedOfNewArticlesExamples.py) for an example of use.

## [v8.7]() (2019-10-16)

**Added**

- added `EventRegistry.getServiceStatus()` method that reports status of the services
- `ComplexQueryArticles` and `ComplexQueryEvents` classes now support in the constructor additional filters like `minSentiment`, `maxSentiment`, `minFacebookShares`, `endSourceRankPercentile`, etc.

**Updated**

- `ReturnInfo` classes (`ArticleInfoFlags`, `ConceptInfoFlags`, ...) were updated. Some obsolete parameters were removed and we have added support for kwdargs to supply some very rarely used parameters directly.
- `TopicPage.getArticles` and `TopicPage.getEvents` methods now support also `**kwargs` so that you can now also provide other available parameters that are less frequently used.

**Removed**

- removed `EventRegistry.suggestCustomConcepts()` and `EventRegistry.getCustomConceptUri()` methods. Not used anymore since we are not supporting anymore the correlation feature.


## [v8.6]() (2019-02-22)

**Added**
- We added sentiment, which can now be used in querying of articles and events. The `QueryArticles`, `QueryArticlesIter`, `QueryEvents`, `QueryEventsIter` constructors now all have additional parameters `minSentiment` and `maxSentiment` that can be used to filter the articles and events. The valid values are between -1 (very negative sentiment) and 1 (very positive sentiment). Value 0 represents neutral sentiment.
- Sentiment was also added as a property in the returned articles and events.

**Updated**

- Analytics: We updated `trainTopicOnTweets()`, `trainTopicClearTopic()` and `trainTopicGetTrainedTopic()` methods in the `Analytics` class.
- `Analytics.annotate()` method now supports passing custom parameters that should be used when annotating the text.
- Changed some defaults in the returned data. When searching articles, we now by default return article image and sentiment.


## [v8.5.1]() (2018-12-12)

**Updated**


Examples.py
from EventRegistry import *

er = EventRegistry(host = "http://eventregistry.org", logging = True)

q = QueryEvents()
q.addConcept(er.getConceptUri("Obama"))                 # get events related to obama
#q.addCategory(er.getCategoryUri("society issues"))      # and are related to issues in society
#q.addNewsSource(er.getNewsSourceUri("bbc"))             # and have been reported by BBC
q.addRequestedResult(RequestEventsUriList())            # return uris of all events
q.addRequestedResult(RequestEventsInfo(page = 0, count = 30, sortBy = "size", sortByAsc = True, conceptLang = "deu", conceptTypes = "person"))   # return event details for first 30 events
q.addRequestedResult(RequestEventsConceptAggr(conceptCount = 5, conceptTypes = ["org", "loc"]))        # compute concept aggregate on the events
res = er.execQuery(q)
obj = createStructFromDict(res)

q = QueryEvents();
q.addLocation(er.getLocationUri("Berlin"))
q.setDateLimit(datetime.date(2014, 4, 16), datetime.date(2014, 4, 28))
q.addRequestedResult(RequestEventsConceptTrends(40, conceptTypes=["person"]))
q.addRequestedResult(RequestEventsCategoryAggr())
q.addRequestedResult(RequestEventsInfo())
res = er.execQuery(q)

q = QueryEvents();
q.addConcept(er.getConceptUri("Obama"))                 # get events related to obama
q.addRequestedResult(RequestEventsConceptGraph(200, linkCount = 500, eventsSampleSize = 2000))
res = er.execQuery(q)

# get info about event with uri "123"
q = QueryEvent("997019");
q.addRequestedResult(RequestEventInfo(["eng", "spa", "slv"]))
q.addRequestedResult(RequestEventArticles(0, 10))        # get 10 articles about the event (any language is ok) that are closest to the center of the event
q.addRequestedResult(RequestEventArticleTrend())
q.addRequestedResult(RequestEventKeywordAggr())
eventRes = er.execQuery(q);

if (eventRes.has_key("articles") and len(eventRes["articles"]) > 0):
    articleUris = [art["uri"] for art in eventRes["articles"]["results"][:5]]  # take only first two articles
    # query info about specific articles
    qa = QueryArticle(articleUris);
    qa.addRequestedResult(RequestArticleInfo())                 # get all info about the specified article
    #qa.addRequestedResult(RequestArticleDuplicatedArticles())   # get info about duplicated articles for the specified two articles
    articleRes = er.execQuery(qa);
            
# search article by uri
q = QueryArticle("247634888");
q.addRequestedResult(RequestArticleInfo())                 # get all info about the specified article
res = er.execQuery(q);

# search article by url
q = QueryArticle.queryByUrl("http://www.bbc.co.uk/news/world-europe-31763789#sa-ns_mchannel%3Drss%26ns_source%3DPublicRSS20-sa");
q.addRequestedResult(RequestArticleInfo())                 # get all info about the specified article
res = er.execQuery(q);


q = QueryArticles();
q.setDateLimit(datetime.date(2014, 4, 16), datetime.date(2014, 4, 28))
#q.addKeyword("apple")
#q.addKeyword("iphone")
q.addConcept(er.getConceptUri("Apple"));
q.addRequestedResult(RequestArticlesInfo(page=0, count = 30, includeArticleDuplicateList = True, includeArticleConcepts = True, includeArticleCategories = True, includeArticleLocation = True, includeArticleImage = True));
res = er.execQuery(q)

obj = createStructFromDict(res);
uris = [article.uri for article in obj.articles.results[:5]]
q = QueryArticle(uris)
q.addRequestedResult(RequestArticleInfo(includeArticleConcepts = True, includeArticleCategories = True, includeArticleLocation = True))
q.addRequestedResult(RequestArticleOriginalArticle())
q.addRequestedResult(RequestArticleDuplicatedArticles())
res = er.execQuery(q);
obj = createStructFromDict(res)

# recent activity
q = QueryEvents()
q.addConcept(er.getConceptUri("Obama"))
q.addRequestedResult(RequestEventsRecentActivity())     # get most recently updated events related to obama
res = er.execQuery(q)
obj = createStructFromDict(res)

# let's say now that a minute has passed
q.clearRequestedResults()
q.addRequestedResult(RequestEventsRecentActivity(lastEventActivityId = obj.recentActivity.events.lastActivityId))
res = er.execQuery(q)
obj2 = createStructFromDict(res)

q = QueryArticles()
q.addConcept(er.getConceptUri("Obama"))
q.addRequestedResult(RequestArticlesRecentActivity())     # get most recently added articles related to obama
res = er.execQuery(q)

recentEvents = er.getRecentEvents()
recentArticles = er.getRecentArticles()

q = QueryArticles();
q.addConcept("topic-page-3")
q.setDateLimit("2014-09-20", "2014-09-29")
q.addRequestedResult(RequestArticlesInfo(page=0, count=10));
res = er.execQuery(q);

# obtain in one call information about events with uris 234, 212 and 423
q = QueryEvent([234, 212, 423]);
q.addRequestedResult(RequestEventArticles(0,200))
res = er.execQuery(q)
obj = createStructFromDict(res)

print er.getRecentStats()

